mcorpc/authz_rego.go

```diff
@@ -0,0 +1,161 @@
+	logger := log.WithFields(logrus.Fields{
+		"authorizer": "regoPolicy",
+		"agent":      agent.Name(),
+		"request":    req.RequestID,
+	})
+
+	authz := &regoPolicy{
+		cfg:     agent.Config,
+		req:     req,
+		agent:   agent,
+		matcher: &regoPolicyPolicy{log: logger},
+		groups:  make(map[string][]string),
+		log:     logger,
+	}
+
+	log.Infof("Agent is %s", agent.Name())
+
+	return authz.authorize()
+}
+
+func (r *regoPolicy) authorize() (bool, error) {
+	policyFile, err := r.lookupPolicyFile()
+	if err != nil {
+		return false, err
+	}
+
+	pol, err := ioutil.ReadFile(policyFile)
+	if err != nil {
+		return false, err
+	}
+
+	module := string(pol)
+
+	query := rego.New(
+		rego.Query("data.choria.authpolicy.allow"),
+		rego.Module("//choria/policies/"+r.agent.meta.Name, module),
+		rego.Input(r.regoInputs()),
+	)
+
+	rs, err := query.Eval(context.Background())
+
+	switch err := err.(type) {
+	case nil:
+		break
+	case ast.Errors:
+		for _, e := range err {
+			r.log.Info("code: ", e.Code)
+			r.log.Info("row: ", e.Location.Row)
+			r.log.Info("filename: ", policyFile)
+		}
+	default:
+		return false, err
+	}
+
+	return rs[0].Expressions[0].Value.(bool), nil
+}
+
+func (r *regoPolicy) defaultPolicyFilenName() string {
+	return r.cfg.Option("plugin.regopolicy.default_name", "default")
+}
+
+func (r *regoPolicyPolicy) Set(caller string, actions string, facts string, classes string, groups map[string][]string) {
+	r.caller = caller
+	r.actions = actions
+	r.facts = facts
+	r.classes = classes
+	r.groups = groups
+}
+
+func (r *regoPolicyPolicy) SetFile(f string) {
+	r.file = f
+}
+
+func (r *regoPolicy) shouldUseDefault() bool {
+	enabled, err := choria.StrToBool(r.cfg.Option("plugin.regopolicy.enable_default", "n"))
+
+	if err != nil {
+		r.log.Debug("Parseing plugin.regopolicy.enable_default failed?:", err)
+		return false
+	}
+
+	return enabled
+}
+
+func (r *regoPolicy) defaultPolicyFileName() string {
+	return r.cfg.Option("plugin.regopolicy.default_name", "default")
+}
+
+func (r *regoPolicy) lookupPolicyFile() (string, error) {
+	dir := filepath.Join(filepath.Dir(r.cfg.ConfigFile), "policies/rego")
+
+	regoPolicy := filepath.Join(dir, r.agent.Name()+".rego")
+
+	r.log.Debugf("Looking up rego policy in %s", regoPolicy)
+	if choria.FileExist((regoPolicy)) {
+		return regoPolicy, nil
+	}
+
+	if r.shouldUseDefault() {
+		defaultPolicy := filepath.Join(dir, r.defaultPolicyFileName()+".rego")
+		if choria.FileExist(defaultPolicy) {
+			return defaultPolicy, nil
+		}
+		return "", fmt.Errorf("No default policy %s found for %s in %s", defaultPolicy, r.agent.Name(), dir)
+
+	}
+
+	return "", fmt.Errorf("No policy found for %s in %s", r.agent.Name(), dir)
+
+}
+
+func (r *regoPolicy) regoInputs() map[string]interface{} {
+	return map[string]interface{}{
+		"agent":      r.req.Agent,
+		"action":     r.req.Action,
+		"callerID":   r.req.CallerID,
+		"collective": r.req.Collective,
+		"ttl":        r.req.TTL,
+		"time":       r.req.Time,
```

Comment:
be nice to get facts, classes and agents in here too

Reply:
facts is `r.agent.ServerInfoSource.Facts()` unmarshal that to a `map[string]interface{}` to add it here. Agents is `ServerInfoSource.KnownAgents()` and classes is `ServerInfoSource.Classes()`

Reply:
Also be good to get `ServerInfoSource.Status()` on here, this will let one write policies that affect provisioning mode for example

Reply:
What dark magic do I need to do to get ServerInfoSource populated in the test suite?

Reply:
Make a mock with mockgen then something like https://github.com/choria-io/mcorpc-agent-provider/blob/12e782656b71291c4ccb6079122030bd75fbd502/mcorpc/golang/rpcutil/rpcutil_test.go#L73